Granular dam break example (#677)
lethe-cfd/lethe: doc/source/examples/dem/granular-dam-break/granular-dam-break.rst

==================================
Granular dam-break
==================================

This example simulates the dam-break flow of a dry granular material using the discrete element method (DEM). More information regarding the DEM parameters are given in the Lethe documentation, i.e. `DEM parameters <../../../parameters/dem/dem.html>`_.

Features
----------------------------------
- Solvers: ``dem_3d``
- Three-dimensional problem
- Moving floating mesh
- Load-balancing

Files used in this example
----------------------------

- Parameters file for the problem (H=20cm): ``examples/dem/3d-dam-break/granular-dam-break.prm``

> - Parameters file for the H=40cm bed: ``examples/dem/3d-dam-break/granular-dam-break-H-40cm.prm``

Reviewer: I think you forgot to add this new prm file

Author: Indeed, good catch :)

==================================
Granular dam-break
==================================

This example simulates the dam-break flow of a dry granular material using the discrete element method (DEM). More information regarding the DEM parameters are given in the Lethe documentation, i.e. `DEM parameters <../../../parameters/dem/dem.html>`_.

Features
----------------------------------
- Solvers: ``dem_3d``
- Three-dimensional problem
- Moving floating mesh
- Load-balancing

Files used in this example
----------------------------

- Parameters file for the problem (H=20cm): ``examples/dem/3d-dam-break/granular-dam-break.prm``

> - Parameters file for the H=40cm bed: ``examples/dem/3d-dam-break/granular-dam-break.prm``

Reviewer: Shouldn't these be two different prm files?

Author: Yup, my mistake :)!

- GMSH geo file: ``examples/dem/3d-dam-break/square.geo``
- GMSH mesh file: ``examples/dem/3d-dam-break/square.msh``

Description of the case
-----------------------

> This example simulates the dam-break flow of a dry powder. First, a packing of particles is generated by inserting particles behind a sluice gate. After 0.5s, the sluice gate rises and the particles are able to flow freely, producing a dam break flow. We compare the results we obtain with those of `Evangelista et al. <https://www.cabdirect.org/cabdirect/abstract/20153319263>`_. To ensure that the simulation remains computationnaly efficient, we will scale-up the sand particles used by Evangelista et al. from a diameter of 1.6mm to 10mm. This reduces their number by a factor of approximately 250x.We first simulate case B1 of Evangelista et al., with a bed length of 60cm and bed height of 20cm. As an extension, we also simulate a bed height of 40cm (case B3).

Reviewer: ```suggestion
This example simulates the dam-break flow of a dry powder. First, a packing of particles is generated by inserting particles behind a sluice gate. After 0.5s, the sluice gate rises and the particles are able to flow freely, producing a dam break flow. We compare the results we obtain with those of `Evangelista et al. <https://www.cabdirect.org/cabdirect/abstract/20153319263>`_. To ensure that the simulation remains computationally efficient, we will scale-up the sand particles used by Evangelista et al. from a diameter of 1.6mm to 10mm. This reduces their number by a factor of approximately 250x. We first simulate case B1 of Evangelista et al., with a bed length of 60cm and bed height of 20cm. As an extension, we also simulate a bed height of 40cm (case B3).
```

Reviewer: I'm not sure if the link to the Evangelista et al. reference is working, but I was not able to access it. Maybe you can add the reference in the end as well? As in the previous DEM examples...

Parameter file
--------------

Mesh
~~~~~

> The domain we simulate is a rectangular box which is :math:`2\times0.4\times0.4` meters which is made using the deal.ii grid generator.  The grid is refined 4 times using the ``set initial refinement`` parameters.

Reviewer: ```suggestion
The domain we simulate is a rectangular box which is :math:`2\times0.4\times0.4` meters and is made using the deal.ii grid generator.  The grid is refined 4 times using the ``set initial refinement`` parameter.
```

.. code-block:: text

  subsection mesh
    set type               = dealii
    set grid type          = subdivided_hyper_rectangle
    set grid arguments     = 5,1,1 : 0,0,0 : 2,0.4,0.4 : false
    set initial refinement = 4
  end

Lagrangian physical properties
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The particles are mono-dispersed with a radius of 0.005 m and a density of 2560 :math:`kg/m3`, respectively. We insert 60k particles to ensure an initial height of 0.2m in order to match the B1 experiment of Evangelista et al. The friction coefficient and the rolling coefficient are the most important parameters in this simulation, since they control the final repose angle of the particles as well as the speed of the progression of the dam-break flow. Since sand is generally non-spherical, we choose relatively high values for the friction and the rolling friction coefficients.

Reviewer: ```suggestion
The particles are mono dispersed with a radius of 0.005 m and a density of 2560 :math:`kg/m^3`, respectively. We insert 60k particles to ensure an initial height of 0.2m in order to match the B1 experiment of Evangelista et al. The friction coefficient and the rolling coefficient are the most important parameters in this simulation, since they control the final repose angle of the particles as well as the speed of the progression of the dam-break flow. Since sand is generally non-spherical, we choose relatively high values for the friction and the rolling friction coefficients.
```

.. code-block:: text

>   subsection lagrangian physical properties
>     set gx                       = 0
>     set gy                       = -9.81
>     set gz                       = 0.0
>     set number of particle types = 1
>     subsection particle type 0
>       set size distribution type            = uniform
>       set diameter                          = 0.01
>       set number                            = 60000
>       set density particles                 = 2560
>       set young modulus particles           = 1e6
>       set poisson ratio particles           = 0.2
>       set restitution coefficient particles = 0.9
>       set friction coefficient particles    = 0.5
>       set rolling friction particles        = 0.3
>   
>     end
>     set young modulus wall           = 1e6
>     set poisson ratio wall           = 0.2
>     set restitution coefficient wall = 0.9
>     set friction coefficient wall    = 0.5
>     set rolling friction wall        = 0.3
>   end

Reviewer: ```suggestion
  subsection lagrangian physical properties
    set gx                       = 0
    set gy                       = -9.81
    set gz                       = 0.0
    set number of particle types = 1
    subsection particle type 0
      set size distribution type            = uniform
      set diameter                          = 0.01
      set number                            = 60000
      set density particles                 = 2560
      set young modulus particles           = 1e6
      set poisson ratio particles           = 0.2
      set restitution coefficient particles = 0.9
      set friction coefficient particles    = 0.5
      set rolling friction particles        = 0.3
    end
    set young modulus wall           = 1e6
    set poisson ratio wall           = 0.2
    set restitution coefficient wall = 0.9
    set friction coefficient wall    = 0.5
    set rolling friction wall        = 0.3
  end
```

Model parameters
~~~~~~~~~~~~~~~~~~~~

> In this example, we use the ``frequent`` load balancing method to ensure that the divion of the domain amongst the various processing cores follow the particles.

Reviewer: ```suggestion
In this example, we use the ``frequent`` load balancing method to ensure that the division of the domain amongst the various processing cores follow the particles.
```

.. code-block:: text

  subsection solid objects
  set number of solids = 1
    subsection solid object 0
      subsection mesh
        set type               = gmsh
        set file name          = square.msh
        set simplex            = true
        set initial refinement = 0
      end
  
      subsection translational velocity
        set Function expression = 0 ; if(t>0.5,if(t<0.7,1,0),0) ; 0
      end
      subsection angular velocity
        set Function expression = 0 ; 0  ; 0
      end
    end
  end

Simulation control
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The simulation runs for 2 seconds of real time. We output the simulation results ever 1k iterations.

Reviewer: ```suggestion
The simulation runs for 2 seconds of real time. We output the simulation results every 1k iterations.
```

Reviewer: Isn't it 500 iterations?

.. code-block:: text

  subsection simulation control
    set time step         = 5e-5
    set time end          = 2
    set log frequency     = 500
    set output frequency  = 500
    set output path       = ./output/
    set output boundaries = true
  end
    

Running the simulation
-----------------------

The simulation is launched using a single input file for the insertion of the particles, the opening of the sluice gate and the granular motion of the particles.

.. code-block:: text

   mpirun -np 8 dem_3d granular-dam-break.prm

.. note::

>  This example needs a simulation time of approximately 10 minutes on 8 processors using an 12th Gen Intel(R) Core(TM) i9-12900K

Reviewer: ```suggestion
 This example needs a simulation time of approximately 10 minutes on 8 processors using an 12th Gen Intel(R) Core(TM) i9-12900K.
```

Results for H=20cm
---------------------

> The following movie displays the dam-break flow

Reviewer: ```suggestion
The following movie displays the dam-break flow:
```

.. raw:: html

    <iframe width="840" height="472"  src="https://www.youtube.com/embed/v32ZqxO2X98" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

The following image compares the final repose angle of the particles with the results of Evangelista et al. We note the good agreement between the two. Evangelista et al. also found experimentally that the particle height profile varied from :math:`0.2m` at :math:`x=0.4m` to :math:`\approx 0.025m` at :math:`x=0.8m` which is in good agreement with what we observe here.

.. image:: images/h20cm_comparison.png
    :alt: Height profile comparison
    :align: center

Results for H=40cm
---------------------

> The following image compares the final granular profile the particles with the results of Evangelista et al (illustrated using a magenta curve). We note the good agreement between the two, except for the fact that a few particles have moved further away (:math:`x>1.1m`) than the one from the experiments of Evangelista et al. This is a consequence of the fact that we have used larger particles to simulate the case.

Reviewer: ```suggestion
The following image compares the final granular profile of the particles with the results of Evangelista et al (illustrated using a magenta curve). We note the good agreement between the two, except for the fact that a few particles have moved further away (:math:`x>1.1m`) than the one from the experiments of Evangelista et al. This is a consequence of the fact that we have used larger particles to simulate the case.
```

.. image:: images/h20cm_comparison.png
    :alt: Height profile comparison
    :align: center

Results for H=40cm
---------------------

The following image compares the final granular profile of the particles with the results of Evangelista et al (illustrated using a magenta curve). We note the good agreement between the two, except for the fact that a few particles have moved further away (:math:`x>1.1m`) than the one from the experiments of Evangelista et al. This is a consequence of the fact that we have used larger particles to simulate the case.

.. image:: images/h40cm_comparison.png
    :alt: Height profile comparison
    :align: center

Possibilities for extension
----------------------------

- Study the impact of the friction and rolling friction coefficients on the dam-break flow.
- Change the height of the bed to 30cm or 50m and try to reproduce the other experiments of Evangelista et al.
- Use smaller particles and reproduce the full experiment of Evagenlista et al.

> eference

Reviewer: ```suggestion
Reference
```
